R/check_tbl_spl_compound_tid.R

```diff
@@ -0,0 +1,81 @@
+#' Check model output data tbl samples contain consistent task ID
+#' combinations for a given compound idx.
+#' @param tbl a tibble/data.frame of the contents of the file being validated. Column types must **all be character**.
+#' @inherit check_tbl_colnames params
+#' @inherit check_tbl_colnames return
+#' @export
+check_tbl_spl_compound_tid <- function(tbl, round_id, file_path, hub_path) {
+  config_tasks <- hubUtils::read_config(hub_path, "tasks")
+
+  if (isFALSE(has_spls_tbl(tbl)) || isFALSE(hubUtils::is_v3_config(config_tasks))) {
+    return(skip_v3_spl_check(file_path))
+  }
+
+  hash_tbl <- spl_hash_tbl(tbl, round_id, config_tasks)
+
+  n_tbl <- dplyr::group_by(hash_tbl, .data$compound_idx) %>%
+    dplyr::summarise(n = dplyr::n_distinct(.data$hash_comp_tid)) %>%
+    dplyr::filter(.data$n > 1L)
+
+  check <- nrow(n_tbl) == 0L
+
+  if (check) {
+    details <- NULL
+    errors <- NULL
+  } else {
+    errors <- comptid_mismatch(
+      n_tbl$compound_idx, hash_tbl, tbl
+    )
+    mismatches <- purrr::map_chr(errors, ~ .x$mismatches) # nolint: object_usage_linter
+    details <- cli::format_inline(
+      "Sample{?s} {.val {mismatches}} d{?oes/o} not match most prevalent ",
+      "task ID combination for {?its/their} compound idx. ",
+      "See {.var errors} attribute for details."
```

Comment:
So the reason I've used "prevalent" has to do with the fact that, within a given submission file, when providing samples, folks do not need to include predictions for all optional values of non-compound task id values. They do however need to include predictions for **the same** optional values across samples (i.e. combinations of values in samples for each compound_idx need to be consistent across all samples).  This is harder to validate than it sounds as there's effectively no exact baseline to compare all samples against. If they are all the same, great, validation passes but if not, how do we decide which samples are incorrect? The way I decided to approach it is to check the frequency of different combinations, take the most frequent combination as the baseline and then declare ones that do not match the baseline as incorrect. This is what I mean by "prevalent". 

Say for example in your test hub that horizons > 10 were optional.  Most sample contain predictions for horizons 1:15 but in a couple of samples, predictions for horizons 1:20 are included.  They are both technically correct value combinations but because the majority of samples only give horizons 1:15, and only a couple deviate from that, the ones that deviate from the most prevalent (i.e. common) task id combination in the sample are flagged as erroneous.

Does the explanation make sense. I think the confusion is justified and was also not sure about the word "prevalent" but it was the best I could come up with. Any suggestions on how to get this message across welcome!!

Reply:
Don't we have a baseline with the information in the associated json?
For example, if we have 6 tasks ids column: `origin_date`, `scenario_id`, `location`, `target`, `horizon`, `age_group` and with a mimimum of 100 samples expected with `"compound_taskid_set": ["origin_date", "scenario_id", "location", "target"]` (I am going to keep the same example all along), is it not expected to have at the minimum for each sample: all the unique horizon and age group possible value, at the minimum? 

|origin_date|scenario_id|location|target|horizon|age_group|output_type_id| compound_idx|
|---|---|---|---|---|---|---| ---|
|2024-04-28| A | US| inc hosp | 1 |65-130|s1|0|
|2024-04-28| A | US| inc hosp | 1 |65-130|s2|0|
...
|2024-04-28| A | US| inc hosp | 1 |65-130|s100|0|
|2024-04-28| A | US| inc hosp | 2 |65-130|s1|0|
|2024-04-28| A | US| inc hosp | 2 |65-130|s2|0|
...
|2024-04-28| A | US| inc hosp | 2 |65-130|s100|0|
...
...
|2024-04-28| A | US| inc hosp | 1 |0-64|s1|0|
|2024-04-28| A | US| inc hosp | 1 |0-64|s2|0|
...
|2024-04-28| A | US| inc hosp | 1 |0-64|s100|0|
...

But a model can submit a file with another level of "compound_task_id_set" (for example: `["origin_date", "location", "target"]`) and it should still be OK because the minimal set will be respected. Here, each sample will have all the unique scenario, horizon and age group possible value.

Reply:
I mean we can check that the minimum is respected and that each sample follow the same compound task id set